turbinia/api/utils.py

```diff
@@ -57,30 +52,97 @@ def create_zip(request_id: str, task_id: str) -> bytes:
-    zip_filename = shutil.make_archive(zip_path, 'zip', request_output_path)
-    data = None
-
-    # Read the zip using a memory-mapped file and return it
-    with open(zip_filename, 'rb') as zip_obj:
-      mm = mmap.mmap(zip_obj.fileno(), 0, access=mmap.ACCESS_READ)
-      data = mm.read()
+class ByteStream:
+  """A writeable in-memory stream used to create tgz files."""
+
+  def __init__(self):
+    """Initialize the object."""
+    self.buffer = io.BytesIO()
+    self.block_size = 4194304  # 4 MB
+    self.offset = 0
+
+  def __enter__(self):
+    return self
+
+  def __exit__(self, exc_type, exc_value, traceback):
+    del exc_type, exc_value, traceback  # Unused.
+    self.close()
+
+  def write(self, data):
+    """Writes data to the buffer and adjusts offset."""
+    self.buffer.write(data)
+    self.offset += len(data)
+
+  def tell(self):
+    """Returns the current byte offset."""
+    return self.offset
+
+  def close(self):
+    """Closes the buffer."""
+    self.buffer.close()
+
+  def pop(self):
+    """Returns the bytes from the buffer.
 
+    Seeks buffer to position 0 and truncates the buffer for re-use.
+    """
+    data = self.buffer.getvalue()
+    self.buffer.seek(0)
+    self.buffer.truncate()
     return data
+
+
+async def create_tarball(output_path: str) -> bytes:
+  """Creates an in-memory TGZ file from output_path contents.
+
+  Partially inspired by the StreamingTarGenerator class from Google's GRR.
+
+  Reference:
+      https://github.com/google/grr/blob/master/grr/core/grr_response_core/lib/utils.py
+
+  Args:
+    output_path (str): The output path of a request or task to archive.
+
+  Yields:
+    bytes: tgz file chunk.
+  """
+  file_paths = []
+  for root, _, filenames in os.walk(output_path):
+    for filename in filenames:
+      file_path = os.path.join(root, filename)
+      log.info(f'Adding {file_path} to tarball.')
+      file_paths.append(file_path)
+
+  with ByteStream() as stream:
+    with tarfile.TarFile.open(fileobj=stream, mode='w:gz',
+                              compresslevel=1) as tar:
+      for file_path in file_paths:
+        tar_info = tar.gettarinfo(name=file_path)
+        tar.addfile(tar_info)
+        # Yield the header for the tarinfo file.
+        yield stream.pop()
+
+        with open(file_path, 'rb') as in_fp:
+          # Read the input file in chunks of stream.block_size bytes.
+          while True:
+            data = in_fp.read(stream.block_size)
+            if len(data) > 0:
+              # Write the data to the buffer.
+              tar.fileobj.write(data)
+              # Yield a compressed file chunk so the client can receive it.
+              yield stream.pop()
+            # Write padding if necessary.
+            if len(data) < stream.block_size:
+              blocks, remainder = divmod(tar_info.size, tarfile.BLOCKSIZE)
+              if remainder > 0:
+                tar.fileobj.write(tarfile.NUL * (tarfile.BLOCKSIZE - remainder))
```

Comment:
Just curious, why do we need to write padding here?

Reply:
This is because each record in the tarball contains blocks of BLOCKSIZE bytes, when getting to the end of the data we're writing to the tar, there would be cases when there is not enough data to fill BLOCKSIZE bytes for the block, so you add some extra bytes up to BLOCKSIZE. 

Ref: https://www.gnu.org/software/tar/manual/html_node/Standard.html

```
The blocks may be blocked for physical I/O operations. Each record of n blocks (where n is set by the ‘--blocking-factor=512-size’ (‘-b 512-size’) option to tar) is written with a single ‘write ()’ operation. On magnetic tapes, the result of such a write is a single record. When writing an archive, the last record of blocks should be written at the full size, with blocks after the zero block containing all zeros. When reading an archive, a reasonable system should properly handle an archive whose last record is shorter than the rest, or which contains garbage records after a zero block.
```